[HUDI-5516] Reduce memory footprint on workload with thousand active partitions

trushev opened the pull request:
This PR significantly reduces memory footprint on workload with thousand active partitions between checkpoints. That workload is relevant with wide checkpoint interval. More specifically, active partition here is a special case of active fileId.
Write client holds map with write handles to create ReplaceHandle between checkpoints. It leads to `OutOfMemoryError` on the workload because write handle is a huge object.
Essentially, it is enough to hold only write path instead of the whole handle.

### Change Logs

1. Released writer in closed create handle. The same approach is used in append and merge handles
1. Replaced `Map<String, HoodieWriteHandle<?, ?, ?, ?>>` with `Map<String, Path>` to reduce memory footprint. All handles are definitely closed by `try-with-resources` with `closeGracefully()`
1. Removed append handle from handle map because it is not used anyway. It reduces memory footprint and fixes potential NPE issue

### Impact

Workload test with `-Xmx2048m`
```SQL
create table source (
    `id` int,
    `data` string
) with (
    'connector' = 'datagen',
    'rows-per-second' = '100',
    'fields.id.kind' = 'sequence',
    'fields.id.start' = '0',
    'fields.id.end' = '3000'
);
create table sink (
    `id` int primary key,
    `data` string,
    `part` string
) partitioned by (`part`) with (
    'connector' = 'hudi',
    'path' = '/tmp/sink',
    'write.batch.size' = '0.001',  -- 1024 bytes
    'write.task.max.size' = '101.001',  -- 101.001MB
    'write.merge.max_memory' = '1'  -- 1024 bytes
);

insert into sink select `id`, `data`, concat('part', cast(`id` as string)) as `part` from source;
```

#### Before

Job failure with `OutOfMemoryError` on partition 550. Heap dump says that handle map holds 1.4 GB

<img width="1300" alt="Снимок экрана 2023-01-09 в 11 13 13" src="https://user-images.githubusercontent.com/42293632/211240084-c1c136d9-b8bf-46c2-9dde-48404fc5f38a.png">

![fix-oom-before](https://user-images.githubusercontent.com/42293632/211239403-d0609d83-e56d-4326-9d5d-64db3c80bce9.png)

#### After 
3000 rows written
![fix-oom-after](https://user-images.githubusercontent.com/42293632/211239453-4898444f-1b76-4132-8bfc-891c716ca92a.png)

### Risk level (write none, low medium or high below)

Low

### Documentation Update

_Describe any necessary documentation update if there is any new feature, config, or user-facing change_

- _The config description must be updated if new configs are added or the default value of the configs are changed_
- _Any new feature or user-facing change requires updating the Hudi website. Please create a Jira ticket, attach the
  ticket number here and follow the [instruction](https://hudi.apache.org/contribute/developer-setup#website) to make
  changes to the website._

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

danny0405 reviewed (commented):
Nice catch, @trushev , curious why the closed handle is also taking huge resource, we may need to figure it out first.

But I still think the change is valid.

trushev commented:
> Nice catch, @trushev , curious why the closed handle is also taking huge resource, we may need to figure it out first.
> 
> But I still think the change is valid.

Thank you for the reviewing

Here is create handle layout with nulled writer. My bad to call it "a huge object even with released writer". Speaking more precise, it is just bigger than we needed. 14 KB against 522 bytes

<img width="878" alt="Снимок экрана 2023-01-10 в 09 31 27" src="https://user-images.githubusercontent.com/42293632/211450420-b6fd1e92-6552-4b9f-af78-2244fa22d41e.png">

trushev commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/StreamWriteFunction.java:
```diff
@@ -449,6 +450,7 @@ private boolean flushBucket(DataBucket bucket) {
     writeStatuses.addAll(writeStatus);
+    writeClient.cleanHandle(bucket.fileID);
     return true;
   }
```
You mean do we need to clean all handles `this.writeClient.cleanHandles()` in `flushRemaining` as we have cleaned handle here?

To be honest, I'm not sure. The handle map would hold unlimited handles. Even lightweight closed handles could exceed heap size. I guess LRU cache solves this problem but not sure about benefits of such approach here

trushev commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/client/HoodieFlinkWriteClient.java:
```diff
@@ -94,7 +94,7 @@
    */
-  private final Map<String, HoodieWriteHandle<?, ?, ?, ?>> bucketToHandles;
+  private final Map<String, MiniBatchHandle> bucketToHandles;
 
```
Thanks, I'll try to rework it

trushev replied:
- Reverted replacing `HoodieWriteHandle` with `MiniBatchHandle`.
- Reverted `FlinkClosedHandle`
- Introduced class `BucketHandles` replacing with `Map<String, HoodieWriteHandle<?, ?, ?, ?>> bucketToHandles` to keep handles and write paths
- Added test `BucketHandles`

Workload benchmark results are the same

danny0405 replied:
We need to clarify whether `HoodieFlinkWriteClient#cleanHandlesGracefully` is still valid. I wrote the logic to want it work like post task killing hook, if a writing task was force killed when it is writing a file, how can we handle this more gracefully.

For most of the case, if the writing task was force killed or canceled, the file writer should had been closed right? I'm skeptical whether it is still needed.

It is related with this issue, because if `cleanHandlesGracefully` is not valid anymore, there is no need to cache any write handles at all.

trushev replied:
Thanks, that is a good point, I'll check it

trushev replied:
I think you are right. There is no need to store writers because they closed by `HoodieConsumer.finish()`. So if we have `List<WriteStatus> writeStatus` then the writer is already closed. We should keep only one `currentWriter` to handle exception during write operation. I'm preparing updated fix now

TengHuo commented:
Hi @trushev 

Nice feature! We are suffering a similar memory exception in our Flink Hudi MOR pipeline. We found a heap OOM exception and abnormal GC activities in task managers.

Task manager GC metrics panel

![tm_gc](https://user-images.githubusercontent.com/7539060/212806036-e3a83720-ba72-42b0-9247-af9ca0913b0c.png)

After checking, we noticed that the size of `CompactionOperation` in memory is unusually big, and it should be caused by `HoodieTableFileSystemView`, because each instance of `HoodieTableFileSystemView` will load all pending compaction plans from the timeline to memory.

This is the part of task manager heap histogram showing the abnormal memory usage caused by `CompactionOperation`.

```log
   9:       2091712       83668480  org.apache.hudi.common.model.CompactionOperation
 479:            27           4752  org.apache.hudi.io.FlinkAppendHandle
 686:            28           2016  org.apache.hudi.common.table.view.HoodieTableFileSystemView
 800:            28           1344  org.apache.hudi.common.table.view.RemoteHoodieTableFileSystemView
 806:            27           1296  org.apache.hudi.table.HoodieFlinkMergeOnReadTable
 954:            27            864  org.apache.hudi.common.table.view.FileSystemViewManager
1064:            28            672  org.apache.hudi.common.table.view.PriorityBasedFileSystemView
```

In the timeline of our pipeline, there was only 1 unfinished compaction plan, which contained 74704 operations, `74704 * 28 = 2091712`.

May I ask if we can lazy load `HoodieTableFileSystemView` in `PriorityBasedFileSystemView` when creating `FlinkAppendHandle`? It can also reduce memory usage for active partitions.

trushev commented:
> May I ask if we can lazy load `HoodieTableFileSystemView` in `PriorityBasedFileSystemView` when creating `FlinkAppendHandle`? It can also reduce memory usage for active partitions.

 @TengHuo thank you for the report. I'll try to reproduce the scenario and consider it here

TengHuo commented:
> > May I ask if we can lazy load `HoodieTableFileSystemView` in `PriorityBasedFileSystemView` when creating `FlinkAppendHandle`? It can also reduce memory usage for active partitions.
> 
> @TengHuo thank you for the report. I'll try to reproduce the scenario and consider it here

Thanks @trushev for your reply. Sorry, my initial idea may not be a proper way to solve this issue. 

The thing which I want to share is that caching write handles could take a lot of memory, because each handle obtains an instance of `HoodieTable`, and there is a `viewManager` in every `HoodieTable`, which will load all pending compaction plans from Hudi timeline when the `FileSystemViewManager#getFileSystemView` is called.

So I'm thinking if it is feasible that we can share one instance of `HoodieTable` when creating a new handle in the method `HoodieFlinkWriteClient#upsert` and other similar methods? As you have implemented this `BucketHandles` for caching all active handles, we can just create one table instance and drop this instance when all handles are removed from the map.

For reproducing this memory issue I got, you need use MOR table, and setup thousands of partitions, as long as one compaction plan generated, you will find each `FlinkAppendHandle` will take a lot of memory because of `CompactionOperation `.

TengHuo commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/client/BucketHandles.java:
```diff
@@ -0,0 +1,107 @@
+public final class BucketHandles {
+  private final Map<String, MiniBatchHandle> fileToHandle;
+  private final Map<String, Path> fileToPath;
+
```
Think we can add a `HoodieTable` attribute here, and share this table object in handles.

I can raise a PR to your branch. What do you think?

trushev replied:
Thank you for helping, I'm reworking the PR now according to https://github.com/apache/hudi/pull/7626#discussion_r1066607740 and `BucketHandles` is going to be removed at all

trushev commented:
> The thing which I want to share is that caching write handles could take a lot of memory, because each handle obtains an instance of `HoodieTable`, and there is a `viewManager` in every `HoodieTable`, which will load all pending compaction plans from Hudi timeline when the `FileSystemViewManager#getFileSystemView` is called.

@TengHuo  I've reworked this PR, we don't cache handles at all now. If it's possible could you pls apply this commit and verify that your scenario is fixed?

trushev commented:
@danny0405 Could you pls take a look again. New solution:
- Replaced `Map<String, HoodieWriteHandle<?, ?, ?, ?>>` with `Map<String, Path>`
- All handles are definitely closed by finally section with `closeGracefully()`

danny0405 reviewed (changes requested):
Thanks for great effort @trushev, can we revert the changes/refactoring for `#performWriteOperation`, it is not related with this issue, and we can address it in another PR.

danny0405 commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/client/HoodieFlinkWriteClient.java:
```diff
@@ -496,4 +474,67 @@ private List<String> getAllExistingFileIds(HoodieFlinkTable<T> table, String par
+    } finally {
+      if (writeHandle != null) {
+        // ensure the writer is closed
+        ((MiniBatchHandle) writeHandle).closeGracefully();
```
Check for `writeHandle != null` seems better?

trushev commented:
> Thanks for great effort @trushev, can we revert the changes/refactoring for `#performWriteOperation`, it is not related with this issue, and we can address it in another PR.
> 
> We can still abstract out some utilities for the `try-finally-close` handling.

Thank you, I replaced the refactoring with simple `try-finally-close` and rebased branch on the latest master

danny0405 reviewed (approved):
+1, nice job!

trushev commented:
@TengHuo 
I tried the following workload with MOR table, 2000 partitions and compaction (checkpoint here triggers compaction)
```java
public class TestPartitionsWorkloadWithCompaction extends TestWriteMergeOnReadWithCompact {
  @Test
  public void write2000partitions() throws Exception {
    int partitionCount = 2000; // aka rowCount
    List<RowData> oneRowPerPartitionData = IntStream.range(0, partitionCount).mapToObj(counter -> TestData.insertRow(
        StringData.fromString("id" + counter),
        StringData.fromString("Name"),
        0,
        TimestampData.fromEpochMillis(counter),
        StringData.fromString("par" + counter))
    ).collect(Collectors.toList());
    conf.setDouble(FlinkOptions.WRITE_BATCH_SIZE, 0.001); // 1024 bytes
    conf.setDouble(FlinkOptions.WRITE_TASK_MAX_SIZE, 101.001); // 101.001MB - 1024 bytes
    conf.setInteger(FlinkOptions.WRITE_MERGE_MAX_MEMORY, 1); // 1024 bytes
    preparePipeline(conf)
        .consume(oneRowPerPartitionData)
        .assertNextEvent()
        .checkpoint(1)
        .checkpointComplete(1)
        .end();
  }
}
```
I guess your problem is fixed by this PR

<img width="714" alt="Снимок экрана 2023-01-19 в 12 39 46" src="https://user-images.githubusercontent.com/42293632/213364715-c0a4c125-7415-4cab-8d94-6916ba85172e.png">

danny0405 commented:
All the Flink related tests are passed so would merge the PR soon ~

TengHuo commented:
> @TengHuo I tried the following workload with MOR table, 2000 partitions and compaction (checkpoint here triggers compaction)

Got it, thanks so much @trushev
